deepcell/applications/application.py

```diff
@@ -0,0 +1,164 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Base class for applications"""
+
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+from deepcell_toolbox.utils import resize, tile_image, untile_image
+
+
+class Application(object):
+    """Application object that takes a model with weights and manages predictions
+    """
+
+    def __init__(self,
+                 model,
+                 model_image_shape=(128, 128, 1),
+                 model_mpp=0.65,
+                 preprocessing_fn=None,
+                 postprocessing_fn=None,
+                 dataset_metadata=None,
+                 model_metadata=None):
+        """Initializes model for application object
+
+        Args:
+            model (tf.model): Tensorflow model with weights loaded
+            model_image_shape (tuple, optional): Shape of input expected by model.
+                Defaults to (128, 128, 1).
+            dataset_metadata (optional): Any input, e.g. str or dict. Defaults to None.
+            model_metadata (optional): Any input, e.g. str or dict. Defaults to None.
+            model_mpp (float, optional): Microns per pixel resolution of training data.
+                Defaults to 0.65.
+            preprocessing_fn (function, optional): Preprocessing function to apply to data
+                prior to prediction. Defaults to None.
+            postprocessing_fn (function, optional): Postprocessing function to apply
+                to data after prediction. Defaults to None.
+                Must accept an input of a list of arrays and then return a single array.
+
+        Raises:
+            ValueError: Preprocessing_fn must be a callable function
+            ValueError: Postprocessing_fn must be a callable function
+        """
+
+        self.model = model
+
+        self.model_image_shape = model_image_shape
+        # Require dimension 1 larger than model_input_shape due to addition of batch dimension
+        self.required_rank = len(self.model_image_shape) + 1
+
+        self.model_mpp = model_mpp
+        self.preprocessing_fn = preprocessing_fn
+        self.postprocessing_fn = postprocessing_fn
+        self.dataset_metadata = dataset_metadata
+        self.model_metadata = model_metadata
+
+        # Test that pre and post processing functions are callable
+        if (self.preprocessing_fn is not None) and (callable(self.preprocessing_fn) is False):
+            raise ValueError('Preprocessing_fn must be a callable function.')
+        if (self.postprocessing_fn is not None) and (callable(self.postprocessing_fn) is False):
+            raise ValueError('Postprocessing_fn must be a callable function.')
+
+    def predict(self, x):
+        raise NotImplementedError
+
+    def _predict_segmentation(self,
```

Comment:
Will all `Application` children have a segmentation prediction? If not, maybe this logic belongs in a child class.  You could have another class `SegmentationApplication` which inherits from `Application`.

Reply:
Yes all application objects should have some form of a predict function. I was reluctant to create another intermediate class because then we would end up with a third layer of inheritance that I don't think adds any clarity.